File: sections/61_icpc_contest_readiness/problems/05_guess_data_structure/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    if (!(cin >> t)) return 0;
    while (t--) {
        int n;
        long long limit;
        cin >> n >> limit;
        vector<long long> values(n);
        for (long long &value : values) cin >> value;

        deque<int> maximum, minimum;
        int left = 0;
        long long answer = 0;
        for (int right = 0; right < n; ++right) {
            while (!maximum.empty() && values[maximum.back()] <= values[right]) maximum.pop_back();
            maximum.push_back(right);
            while (!minimum.empty() && values[minimum.back()] >= values[right]) minimum.pop_back();
            minimum.push_back(right);

            while (values[maximum.front()] - values[minimum.front()] > limit) {
                if (maximum.front() == left) maximum.pop_front();
                if (minimum.front() == left) minimum.pop_front();
                ++left;
            }
            answer += right - left + 1;
        }
        cout << answer << '\n';
    }
    return 0;
}

File: sections/61_icpc_contest_readiness/problems/05_guess_data_structure/solution.py
from collections import deque
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    index = 0
    tests = data[index]
    index += 1
    out: list[str] = []
    for _ in range(tests):
        n = data[index]
        limit = data[index + 1]
        index += 2
        values = data[index:index + n]
        index += n

        maximum: deque[int] = deque()
        minimum: deque[int] = deque()
        left = 0
        answer = 0
        for right, value in enumerate(values):
            while maximum and values[maximum[-1]] <= value:
                maximum.pop()
            maximum.append(right)
            while minimum and values[minimum[-1]] >= value:
                minimum.pop()
            minimum.append(right)

            while values[maximum[0]] - values[minimum[0]] > limit:
                if maximum[0] == left:
                    maximum.popleft()
                if minimum[0] == left:
                    minimum.popleft()
                left += 1
            answer += right - left + 1
        out.append(str(answer))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()
